src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+
+    case PredicateCondition::Between: {
+      // `a BETWEEN b AND c` is evaluated by transforming it to `a >= b AND a <= c` instead of evaluating it with a
+      // dedicated algorithm. This is because three expression data types (from three arguments) generate many type
+      // combinations and thus lengthen compile time and increase binary size notably.
+
+      const auto& between_expression = static_cast<const BetweenExpression&>(predicate_expression);
+      const auto gte_expression = greater_than_equals(between_expression.value(), between_expression.lower_bound());
+      const auto lte_expression = less_than_equals(between_expression.value(), between_expression.upper_bound());
+
+      const auto gte_lte_expression = and_(gte_expression, lte_expression);
+
+      return evaluate_expression_to_result<int32_t>(*gte_lte_expression);
+    }
+
+    case PredicateCondition::In:
+      return _evaluate_in_expression<int32_t>(static_cast<const InExpression&>(predicate_expression));
+
+    case PredicateCondition::Like:
+    case PredicateCondition::NotLike:
+      return _evaluate_like_expression<int32_t>(static_cast<const BinaryPredicateExpression&>(predicate_expression));
+
+    case PredicateCondition::IsNull:
+    case PredicateCondition::IsNotNull:
+      return _evaluate_is_null_expression<int32_t>(static_cast<const IsNullExpression&>(predicate_expression));
+  }
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_predicate_expression(
+    const AbstractPredicateExpression& expression) {
+  Fail("Can only evaluate predicates to int32_t (aka bool)");
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_column_expression(
+    const PQPColumnExpression& column_expression) {
+  Assert(_chunk, "Cannot access Columns in this Expression as it doesn't operate on a Table/Chunk");
+
+  const auto& column = *_chunk->get_column(column_expression.column_id);
+  Assert(column.data_type() == data_type_from_type<R>(), "Can't evaluate column to different type");
+
+  _ensure_column_materialization(column_expression.column_id);
+  return std::static_pointer_cast<ExpressionResult<R>>(_column_materializations[column_expression.column_id]);
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_case_expression(
+    const CaseExpression& case_expression) {
+  const auto when = evaluate_expression_to_result<int32_t>(*case_expression.when());
+
+  std::shared_ptr<ExpressionResult<R>> result;
+
+  _resolve_to_expression_results(
+      *case_expression.then(), *case_expression.else_(), [&](const auto& then_result, const auto& else_result) {
+        using ThenResultType = typename std::decay_t<decltype(then_result)>::Type;
+        using ElseResultType = typename std::decay_t<decltype(else_result)>::Type;
+
+        const auto result_size = _result_size(when->size(), then_result.size(), else_result.size());
```

Comment:
Doesn't the result_size solely depend on the number of input rows?

Reply:
After reading more of the code: I don't get the whole `_result_size` thing. Could you please explain that a bit?

Reply:
As per the comment in `_result_size`: `// If any operand is empty (that's the case IFF it is an empty column) the result of the expression has no rows`

`_result_size()` covers the following scenarios:

* Column-involving expression evaluation on an empty Chunk should give you zero rows. So `a + 5` should be empty on an empty Chunk.
* If the Chunk is NOT empty, Literal-and-Column involving expression evaluation should give you one result per row, so `a + 5` should give you one value for each element in a.
* Non-column involving expressions should give you one result value, no matter whether there is a non-empty Chunk involved or not. So `5+3` always gives you one result element: `8`

Reply:
Got it, thanks. Maybe you could copy the list of scenarios to the doc?